### PR TITLE
SONARJS-709 Use live variable analysis to avoid execution duplicated paths

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java [hunk 1]
```diff
@@ -97,215 +91,46 @@ private void checkFunction(FunctionTree functionTree) {
...
 
-  private void buildUsagesAndLivenesses(ControlFlowGraph cfg, Usages usages, Set<BlockLiveness> livenesses) {
-    Map<CfgBlock, BlockLiveness> livenessPerBlock = new HashMap<>();
-    for (CfgBlock block : cfg.blocks()) {
-      BlockLiveness blockLiveness = new BlockLiveness(block, usages);
-      livenessPerBlock.put(block, blockLiveness);
-      livenesses.add(blockLiveness);
-    }
+          if (isWrite(usage)) {
+            if (!live.contains(symbol) && !usages.hasUsagesInNestedFunctions(symbol) && !usages.neverReadSymbols().contains(symbol)) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/DeadStoreCheck.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java [hunk 2]
```diff
@@ -192,4 +194,14 @@ public int hashCode() {
   public SymbolicValue peekStack() {
     return stack.peek();
   }
+
+  public ProgramState cleanWithLva(Set<Symbol> liveOutSymbols) {
+    Map<Symbol, SymbolicValue> newValues = new HashMap<>();
+    values.forEach((symbol, value) -> {
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java]: `com.google.common.collect.Maps.filterKeys(...)` may help.

File: javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java [hunk 3]
```diff
@@ -192,4 +194,14 @@ public int hashCode() {
   public SymbolicValue peekStack() {
     return stack.peek();
   }
+
+  public ProgramState cleanWithLva(Set<Symbol> liveOutSymbols) {
```
Comment (pynicolas) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java]: @vilchik-elena Suggested name: `removeSymbols(symbolsToRemove)`

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java]: @pynicolas thanks! good suggestion (only thing it's not `symbolsToRemove` but `symbolsToKeep` :))

File: javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java (hunk 2 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java]: @pynicolas thanks! fixed 861d2cf8313d0cd03f79f684882bb627924835c0

File: javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java (hunk 3 above)
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/ProgramState.java]: fixed 861d2cf8313d0cd03f79f684882bb627924835c0

**Pull request merged**